tests/std/tests/GH_001858_iostream_exception/test.cpp

```diff
@@ -209,6 +227,204 @@ void test_ostream_exceptions() {
+STATIC_ASSERT(is_nothrow_std_swappable<istringstream>);
+STATIC_ASSERT(is_nothrow_std_swappable<wistringstream>);
+STATIC_ASSERT(is_nothrow_std_swappable<ostringstream>);
+STATIC_ASSERT(is_nothrow_std_swappable<wostringstream>);
+STATIC_ASSERT(is_nothrow_std_swappable<stringstream>);
+STATIC_ASSERT(is_nothrow_std_swappable<wstringstream>);
+
+STATIC_ASSERT(has_nothrow_rdbuf<istringstream>);
+STATIC_ASSERT(has_nothrow_rdbuf<wistringstream>);
+STATIC_ASSERT(has_nothrow_rdbuf<ostringstream>);
+STATIC_ASSERT(has_nothrow_rdbuf<wostringstream>);
+STATIC_ASSERT(has_nothrow_rdbuf<stringstream>);
+STATIC_ASSERT(has_nothrow_rdbuf<wstringstream>);
+
+STATIC_ASSERT(is_nothrow_move_assignable_v<strstreambuf>);
+STATIC_ASSERT(is_nothrow_move_assignable_v<istrstream>);
+STATIC_ASSERT(is_nothrow_move_assignable_v<ostrstream>);
+STATIC_ASSERT(is_nothrow_move_assignable_v<strstream>);
+
+STATIC_ASSERT(noexcept(declval<strstreambuf&>().clear()));
+
+STATIC_ASSERT(noexcept(declval<strstreambuf&>().freeze()));
+STATIC_ASSERT(noexcept(declval<ostrstream&>().freeze()));
+STATIC_ASSERT(noexcept(declval<strstream&>().freeze()));
+
+STATIC_ASSERT(noexcept(declval<strstreambuf&>().str()));
+STATIC_ASSERT(noexcept(declval<istrstream&>().str()));
+STATIC_ASSERT(noexcept(declval<ostrstream&>().str()));
+STATIC_ASSERT(noexcept(declval<strstream&>().str()));
+
+STATIC_ASSERT(noexcept(declval<strstreambuf&>().pcount()));
+STATIC_ASSERT(noexcept(declval<ostrstream&>().pcount()));
+STATIC_ASSERT(noexcept(declval<strstream&>().pcount()));
+
+#if !_HAS_CXX23 // non-Standard overloads
+STATIC_ASSERT(is_nothrow_std_swappable<strstreambuf>);
+STATIC_ASSERT(is_nothrow_std_swappable<istrstream>);
+STATIC_ASSERT(is_nothrow_std_swappable<ostrstream>);
+STATIC_ASSERT(is_nothrow_std_swappable<strstream>);
+#endif // !_HAS_CXX23
+
+#if _HAS_CXX17
+static_assert(is_nothrow_swappable_v<filebuf>);
+static_assert(is_nothrow_swappable_v<wfilebuf>);
+static_assert(is_nothrow_swappable_v<ifstream>);
+static_assert(is_nothrow_swappable_v<wifstream>);
+static_assert(is_nothrow_swappable_v<ofstream>);
+static_assert(is_nothrow_swappable_v<wofstream>);
+static_assert(is_nothrow_swappable_v<fstream>);
+static_assert(is_nothrow_swappable_v<wfstream>);
+
+static_assert(is_nothrow_swappable_v<stringbuf>);
+static_assert(is_nothrow_swappable_v<wstringbuf>);
+static_assert(is_nothrow_swappable_v<istringstream>);
+static_assert(is_nothrow_swappable_v<wistringstream>);
+static_assert(is_nothrow_swappable_v<ostringstream>);
+static_assert(is_nothrow_swappable_v<wostringstream>);
+static_assert(is_nothrow_swappable_v<stringstream>);
+static_assert(is_nothrow_swappable_v<wstringstream>);
+#endif // _HAS_CXX17
+
+#if _HAS_CXX20
+static_assert(is_nothrow_swappable_v<syncbuf>);
+static_assert(is_nothrow_swappable_v<wsyncbuf>);
+
+#ifndef __EDG__
+static_assert(is_nothrow_move_assignable_v<osyncstream>); // mandatory
+static_assert(is_nothrow_move_assignable_v<wosyncstream>); // mandatory
+#endif // __EDG__

```

Comment:
At first, I was going to say that we should report this compiler bug instead of silently working around it forever.

However, something is very strange here. WG21-N4928 [syncstream.osyncstream.overview] indeed depicts `basic_osyncstream& operator=(basic_osyncstream&&) noexcept;`. However, it also depicts an exposition-only `syncbuf_type sb;` and [syncstream.syncbuf.assign] depicts `basic_syncbuf& operator=(basic_syncbuf&& rhs);` which is **not** `noexcept` (it calls `emit()`).

I think there's an LWG issue here, and there might also be a compiler bug, but we shouldn't add these lines to the test until we clearly understand what's happening.